### ISSUE TITLE
Disposer is not invoked when the resource construction contains two nested then's

### ISSUE THREAD

**Issue by hvrauhal** [petkaantonov/bluebird#433]
In the following code, resource 2 is not disposed at all with bluebird versions > 2.4.3:

```javascript
var BPromise = require('bluebird');

function workingGetResource() {
  return BPromise.resolve(1).then(function (x) {
    console.log('Providing resource', x);
    return BPromise.resolve(x)
  }).disposer(function (y) {
    console.log('Disposing resource', y);
  })
}

function failingGetResource() {
  return BPromise.resolve(2).then(function (x) {
    return BPromise.resolve(x).then(function (y) {
      console.log('Providing resource', y);
      return y
    })
  }).disposer(function (z) {
    console.log('Disposing resource', z);
  })
}

BPromise.using(workingGetResource(), function (r) {
  console.log('Using resource', r)
}).then(function () {
  return BPromise.using(failingGetResource(), function (r) {
    console.log('Using resource ' + r + ' that is only disposed when bluebird <= 2.4.2')
  })
});
```

**Comment by hvrauhal** [comment]
Here's a patch with a failing test:
https://github.com/hvrauhal/bluebird/commit/4638b818c5d74cd5d752a8abae1b791538554447

**Comment by petkaantonov** [comment]
Thanks, fixed with the 2.6.3 release
